packages/story-editor/src/components/form/mediaUploadButton.js

```diff
@@ -0,0 +1,156 @@
+import useFFmpeg from '../../app/media/utils/useFFmpeg';
+
+function MediaUploadButton({ onInsert, renderButton, buttonInsertText }) {
+  const {
+    allowedTranscodableMimeTypes,
+    allowedFileTypes,
+    allowedMimeTypes: {
+      image: allowedImageMimeTypes,
+      video: allowedVideoMimeTypes,
+    },
+    MediaUpload,
+  } = useConfig();
+
+  const {
+    canTranscodeResource,
+    resetWithFetch,
+    postProcessingResource,
+    optimizeVideo,
+    optimizeGif,
+  } = useLocalMedia(
+    ({
+      state: { canTranscodeResource },
+      actions: {
+        resetWithFetch,
+        postProcessingResource,
+        optimizeVideo,
+        optimizeGif,
+      },
+    }) => {
+      return {
+        canTranscodeResource,
+        resetWithFetch,
+        postProcessingResource,
+        optimizeVideo,
+        optimizeGif,
+      };
+    }
+  );
+  const { isTranscodingEnabled } = useFFmpeg();
+  const { showSnackbar } = useSnackbar();
+
+  const allowedMimeTypes = useMemo(() => {
+    if (isTranscodingEnabled) {
+      return [
+        ...allowedTranscodableMimeTypes,
+        ...allowedImageMimeTypes,
+        ...allowedVideoMimeTypes,
+      ];
+    }
+    return [...allowedImageMimeTypes, ...allowedVideoMimeTypes];
+  }, [
+    allowedImageMimeTypes,
+    allowedVideoMimeTypes,
+    isTranscodingEnabled,
+    allowedTranscodableMimeTypes,
+  ]);
+
+  const transcodableMimeTypes = useMemo(() => {
+    return allowedTranscodableMimeTypes.filter(
+      (x) => !allowedVideoMimeTypes.includes(x)
+    );
+  }, [allowedTranscodableMimeTypes, allowedVideoMimeTypes]);
+
+  let onSelectErrorMessage = __(
+    'No file types are currently supported.',
+    'web-stories'
+  );
+  if (allowedFileTypes.length) {
+    onSelectErrorMessage = sprintf(
+      /* translators: %s: list of allowed file types. */
+      __('Please choose only %s to insert into page.', 'web-stories'),
+      translateToExclusiveList(allowedFileTypes)
+    );
+  }
+
+  const onSelect = useCallback(
+    (resource) => {
+      try {
+        if (isTranscodingEnabled && canTranscodeResource(resource)) {
+          if (transcodableMimeTypes.includes(resource.mimeType)) {
+            optimizeVideo({ resource });
+          }
+
+          if (resource.mimeType === 'image/gif') {
+            optimizeGif({ resource });
+          }
+        }
+        onInsert(resource);
+
+        postProcessingResource(resource);
+      } catch (e) {
+        showSnackbar({
+          message: e.message,
+          dismissible: true,
+        });
+      }
+    },
+    [
+      isTranscodingEnabled,
+      canTranscodeResource,
+      onInsert,
+      optimizeGif,
+      optimizeVideo,
+      postProcessingResource,
+      showSnackbar,
+      transcodableMimeTypes,
+    ]
+  );
+
+  return (
+    <MediaUpload
+      onSelect={onSelect}
+      onSelectErrorMessage={onSelectErrorMessage}
+      onClose={resetWithFetch}
+      type={allowedMimeTypes}
+      render={renderButton}
```

Comment:
Not completely sure why this is rendered as a callback and not just a (child) component, but I know that's what it also did before, so I guess that's fine.